src/coreclr/jit/importer.cpp

```diff
@@ -10100,22 +10100,38 @@ void Compiler::impImportBlockCode(BasicBlock* block)
                 }
                 else
                 {
+                    // Don't optimize, just call the helper and be done with it
                     JITDUMP("\n Importing %s as helper call because %s\n", opcode == CEE_UNBOX ? "UNBOX" : "UNBOX.ANY",
                             canExpandInline ? "want smaller code or faster jitting" : "inline expansion not legal");
 
-                    // Don't optimize, just call the helper and be done with it
-                    op1 = gtNewHelperCallNode(helper,
-                                              (var_types)((helper == CORINFO_HELP_UNBOX) ? TYP_BYREF : TYP_STRUCT), op2,
-                                              op1);
-                    if (op1->gtType == TYP_STRUCT)
+                    if (helper == CORINFO_HELP_UNBOX)
                     {
-                        op1->AsCall()->gtRetClsHnd = resolvedToken.hClass;
+                        op1 = gtNewHelperCallNode(helper, TYP_BYREF, op2, op1);
+                    }
+                    else
+                    {
+                        assert(helper == CORINFO_HELP_UNBOX_NULLABLE);
+
+                        // We're going to emit the following sequence of IR:
+                        //
+                        //  Nullable<T> result;
+                        //  void CORINFO_HELP_UNBOX_NULLABLE(&result, unboxCls, obj);
+                        //  push result;
+                        //
+                        unsigned resultTmp = lvaGrabTemp(true DEBUGARG("Nullable<T> tmp"));
+                        lvaSetStruct(resultTmp, resolvedToken.hClass, false);
+
+                        GenTreeLclFld* resultAddr = gtNewLclAddrNode(resultTmp, 0);
+                        // NOTE: it's fine for op2 to be evaluated before op1
+                        GenTreeCall* helperCall = gtNewHelperCallNode(helper, TYP_VOID, resultAddr, op2, op1);
+                        impAppendTree(helperCall, CHECK_SPILL_ALL, impCurStmtDI);
+                        op1 = gtNewLclvNode(resultTmp, TYP_STRUCT);
                     }
```

Comment:
Can you post diffs of this change alone?

Reply:
Sure will post here, I expect a few regressions, but this path is expected to be on a cold path with https://github.com/dotnet/runtime/issues/104954

Reply:
Jit-diff (linux-x64): https://github.com/MihuBot/runtime-utils/issues/535